### PR TITLE
[SHELL32] Enable Network Connection Rename CORE-18750

### PR DESCRIPTION
**Pull request by zecarlos1957**
[SHELL32] Enable Network Connection Rename 

Add Network Connection pidl type 0x99
Value hard coded at netshell/enumlist.cpp in function ILCreateNetConnectionItem(...)

ms say:
ILIsEqual performs a binary comparison of the item data.

JIRA issue: [CORE-18750](https://jira.reactos.org/browse/CORE-18750)

### CODE REVIEW

File: dll/win32/shell32/wine/pidl.c [hunk 1]
```diff
@@ -527,6 +528,21 @@ BOOL _ILHACKCompareSimpleIds(LPCITEMIDLIST pidltemp1, LPCITEMIDLIST pidltemp2)
 
     return TRUE;
 }
+#else
+BOOL _ILCompareSimpleIDs(LPCITEMIDLIST pidl1, LPCITEMIDLIST pidl2)
```
Comment (SergeGautherie) [inline dll/win32/shell32/wine/pidl.c:532]: This function can be rewritten to be a single return expression.

File: dll/win32/shell32/debughlp.cpp [hunk 2]
```diff
@@ -379,6 +379,7 @@ BOOL pcheck( LPCITEMIDLIST pidl )
                 case PT_YAGUID:
                 case PT_IESPECIAL2:
                 case PT_SHARE:
+                case 0x99:      /* Network Connection pidl type */
```
Comment (SergeGautherie) [inline dll/win32/shell32/debughlp.cpp:382]: Make it a `#define`?

File: dll/win32/shell32/wine/pidl.c [hunk 3]
```diff
@@ -554,7 +570,11 @@ BOOL WINAPI ILIsEqual(LPCITEMIDLIST pidl1, LPCITEMIDLIST pidl2)
 
     while (pidltemp1->mkid.cb && pidltemp2->mkid.cb)
     {
+#ifdef __REACTOS__
+        if (!_ILCompareSimpleIDs(pidltemp1, pidltemp2))
```
Comment (SergeGautherie) [inline dll/win32/shell32/wine/pidl.c:574]: `#define` an alias, instead.

File: dll/win32/shell32/wine/pidl.c (hunk 1 above)
Comment (learn-more) [reply dll/win32/shell32/wine/pidl.c:532]: It could, but that does not make it better or more readable.

Comment (SergeGautherie) [reply dll/win32/shell32/wine/pidl.c:532]: > not [...] more readable

I have a different point of view:
```
if (false cond) return false
else if (false cond) return false
return true
```
looks quite bloated to me, even more with the additional braces.

File: dll/win32/shell32/debughlp.cpp (hunk 2 above)
Comment (learn-more) [reply dll/win32/shell32/debughlp.cpp:382]: Netshell itself does not even have a define for this, so this is following the existing style.

File: dll/win32/shell32/wine/pidl.c (hunk 3 above)
Comment (learn-more) [reply dll/win32/shell32/wine/pidl.c:574]: I don't like this suggestion.

**Review by learn-more — APPROVED** [review]
lgtm

File: dll/win32/shell32/wine/pidl.c [hunk 4]
```diff
@@ -476,6 +476,7 @@ LPITEMIDLIST WINAPI ILGlobalClone(LPCITEMIDLIST pidl)
     return newpidl;
 }
 
+#ifndef __REACTOS__
 BOOL _ILHACKCompareSimpleIds(LPCITEMIDLIST pidltemp1, LPCITEMIDLIST pidltemp2)
```
Comment (HBelusca) [inline dll/win32/shell32/wine/pidl.c:480]: This "HACKCompare..." function was introduced there: https://github.com/reactos/reactos/commit/2af89ef8f01423c913f47df6e321cc1cec07b3bc
by @yagoulas . So it doesn't need to be "ifndef __REACTOS__"'ed.

**Comment by yagoulas** [comment]
I'm not really confident that we can remove all these special cases for the pidls that are internal to shell32. If I remember correctly there were applications that passed pidls for CFSFolder with forms from older versions of windows (like the ones written in a shortcut file). Some tests may do crazy things like that.

 However I'm totally for adding an RtlEqualMemory call in the last else clause that now returns just FALSE.

**Comment by learn-more** [comment]
> I'm not really confident that we can remove all these special cases for the pidls that are internal to shell32. If I remember correctly there were applications that passed pidls for CFSFolder with forms from older versions of windows (like the ones written in a shortcut file). Some tests may do crazy things like that.
> 
> However I'm totally for adding an RtlEqualMemory call in the last else clause that now returns just FALSE.

Any chance you could dig up an issue that was fixed by those special cases?

**Comment by KRosUser2** [comment]
@yagoulas @learn-more : Do you consider a change is finally stil needed here or is the PR "approved" ?

**Comment by katahiromz** [comment]
Cc @reactos/lang-russian

**Comment by tkreuzer** [comment]
Please provide testbot results.

**Review by Doug-Lyons — APPROVED** [review]
LGTM.

**Comment by zecarlos1957** [comment]
If we don't interpret any part of the pidls and leave only the binary comparison, when we delete a file (only in delete) the file is only deleted when we close the browser window.
Deleting a file triggers the error (win32ss/user/ntuser/msgqueue.c:2398): MQ Cleanup Post Messages E134E4F8. Having as a side effect a failure in the logic of the _ILCompareIds function, which always returns FALSE.

If I remember correctly, if we keep the GUIID test beyond the binary comparison solves this problem, said error is not triggered.

Are there advantages in this perspective?

**Comment by zecarlos1957** [comment]
If we dont interpret any part of the pidls and we let just the binary comparation, when we delete a file (only on delete) the file is delected just when, and if we close the browser window.
Deleting a file, triggers the Error (win32ss/user/ntuser/msgqueue.c:2398) err: MQ Cleanup Post Messages E134E4F8.
Having as a side effect failure in the logic of the _ILCompareIds function , which always returns FALSE

**Comment by zecarlos1957** [comment]
If we do not interpret any part of the pidls and we let just the binary comparison, when we delete a file (only on delete) the file is deleted just when, and if we close the browser window.

**Comment by zecarlos1957** [comment]
_ILCompareIDs definitely doesn't work without the special cases in place.
The present proposal really brings an improvement because it enables the renaming of the shell item "Network Connections" and simultaneously can delete a file without problems, which does not happen in the previous proposal.

**Review by learn-more — APPROVED** [review]

**Review by yagoulas — APPROVED** [review]

File: dll/win32/shell32/wine/pidl.c [hunk 5]
```diff
@@ -522,7 +522,14 @@ BOOL _ILHACKCompareSimpleIds(LPCITEMIDLIST pidltemp1, LPCITEMIDLIST pidltemp2)
     }
     else
     {
-        return FALSE;
+        if (pidltemp1->mkid.cb != pidltemp2->mkid.cb)
+        {
+            return FALSE;
+        }
+        else if (!RtlEqualMemory((BYTE*)&pidltemp1->mkid, (BYTE*)&pidltemp2->mkid, pidltemp1->mkid.cb))
```
Comment (HBelusca) [inline dll/win32/shell32/wine/pidl.c:529]:
```suggestion
        if ((pidltemp1->mkid.cb != pidltemp2->mkid.cb) ||
            !RtlEqualMemory((BYTE*)&pidltemp1->mkid, (BYTE*)&pidltemp2->mkid, pidltemp1->mkid.cb))
```

**Comment by HBelusca** [comment]
@GeoB99 Why "no squash merge" ?

File: dll/win32/shell32/wine/pidl.c (hunk 1 above)
Comment (SergeGautherie) [reply dll/win32/shell32/wine/pidl.c]: Eventually improved/done.

File: dll/win32/shell32/wine/pidl.c (hunk 3 above)
Comment (SergeGautherie) [reply dll/win32/shell32/wine/pidl.c]: Renaming was eventually reverted.

**Pull request merged**